### DISCUSSION TITLE
Possible to add RAG and small GUI?

### DISCUSSION THREAD

**Discussion by fxtoofaan** [crewAIInc/crewAI#1101] [Feature Requests]
Would love to see some kind of RAG and UI implemented. Thank you

**Comment by hugofreire** [comment]
RAG will be awesome

**Comment by fxtoofaan** [comment]
local RAG and local ollama llm model(s) example code please :)

**Comment by joaomdmoura** [comment]
Nice, okay I hear you, I'll look into a specific simple RAG default tool maybe and add more ollama examples

**Comment by okoliechykwuka** [comment]
Thanks, @joaomdmoura , I noticed you implemented RAG in the financial analyst example. 

https://youtu.be/e0Uj4yWdaAg

**Comment by mindwellsolutions** [comment]
@joaomdmoura Thank you for working on adding RAG, this will be huge. The ability for CrewAI to read through a whole folder of  documents, pdfs (potentially spreadsheets, ebooks, datasets) as its research source, in place of DuckDuckGoSearch using the internet for research. This would allow people to point the agent towards larger bodies of documents, knowledge, pdfs and ebooks, to help agents tap into local sources for research and responses. This could be incredibly powerful :)

Support for API Embedding to VectorDBs (ChromaDB or Pinecone) using Gemini/ChatGPT API and Local Embedding (Like Instructor-XL) would be great. Although API is likely the preferred option for ease of scalability and speed.

**Reply by ceramicv** [reply]
Especially useful would be to be able to sense an update of the RAG environment automatically so that you don't have to re-read everything back in if you just add a new document. This is maybe outside of CrewAI itself as all it does is read the RAG database but I haven't seen an elegant way to deal with adds, deleted, modifications to the RAG database so a best practice recommendation would be great.

**Comment by ewebgh33** [comment]
+1

I would really, really like to be able to 
a) give agents bodies of specific knowledge via RAG

And 
b) also give them examples of their expected output as docs they can reference.

Maybe (b) can be achieved in prompt, but this would be clunky I feel.
Or (b) maybe if we're using ChatGPT4.x, we can feed it a document and hope it understands the format enough to mimic on output.

But definitely (a) giving them specific knowledge would be the next logical step to empowering the agents to do different jobs?

**Comment by joaomdmoura** [comment]
This is getting prioritized to be worked on next

**Comment by ewebgh33** [comment]
Excellent news! Very excited. 

I think this would really supercharge crewAI and give it the best of both worlds that Autogen (agents) and taskweaver (code-first) have. One of the main things I would love to try is mixing general knowledge with coding ability, and with Autogen/Taskweaver I feel you get one or the other - not both!

**Comment by WismutHansen** [comment]
Would this help with implementing RAG: https://github.com/KillianLucas/aifs

**Comment by scaruslooner** [comment]
excited

**Comment by slavakurilyak** [comment]
[AI File System](https://github.com/KillianLucas/aifs) can be used as a tool for crewAI that would unlock RAG-pipelines for local search (file or folder search)

**Comment by slavakurilyak** [comment]
[RAG](https://github.com/joaomdmoura/crewai-tools/tree/main/crewai_tools/tools/rag) is now part of the [crewai-tools](https://github.com/joaomdmoura/crewai-tools) library

**Comment by slavakurilyak** [comment]
Excited to see the new UI

**Comment by bhargav-11** [comment]
bump!!!

**Comment by airtonix** [comment]
don't waste time on a UI. it's orthogonal to the purpose of this library.

**Comment by joaomdmoura** [comment]
Yup UI us coming, we kept on pushing it to build the backbones of the lib but we will also need it ourselves, so expected it sooner than later

**Comment by joaomdmoura** [comment]
https://github.com/joaomdmoura/crewAI/issues/18#issuecomment-1955750679
About RAG ^

**Comment by magedhelmy1** [comment]
Dear @joaomdmoura, sorry to bother you on this. What is the ETA? 
PS, keep the good work up!

**Comment by joaomdmoura** [comment]
UI is in the works and scheduled to drop in the next 30d :)
Rag is now supported over tools :D
--
João Moura
@joaomdmoura <http://twitter.com/joaomdmoura>


Em ter., 30 de abr. de 2024 às 18:19, Maged Helmy ***@***.***>
escreveu:

**Reply by shawnholt** [reply]
Please consider teaming up with Langflow, Flowise or similar.

**Comment by shawnholt** [comment]
A lot of people have been drawn to autogen because it's much more comfortable to work with a GUI , rather than code. I see a lot of abandoned attempts and "coming soon", but based on this thread alone it does not seem like it is a priority, please help us understand where it fits into the roadmap. Thank you!

**Comment by YohanReddy** [comment]
> UI is in the works and scheduled to drop in the next 30d :) Rag is now supported over tools :D -- João Moura @joaomdmoura <http://twitter.com/joaomdmoura> Em ter., 30 de abr. de 2024 às 18:19, Maged Helmy ***@***.***> escreveu:
> […](#)
> Dear @joaomdmoura <https://github.com/joaomdmoura>, sorry to bother you on this. What is the ETA? PS, keep the good work up! — Reply to this email directly, view it on GitHub <[#18 (comment)](https://github.com/joaomdmoura/crewAI/issues/18#issuecomment-2087360532)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAFC3N6JZQS4MDOVERYFNS3ZAADGXAVCNFSM6AAAAABBE2PH6CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOBXGM3DANJTGI> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Any update on the UI? Can't wait to get hands-on

**Comment by cbarkinozer** [comment]
Is there any news or an ETA from UI?

**Reply by beevelop** [reply]
Langflow announced support for CrewAI recently which could be a feasible interim option: https://github.com/langflow-ai/langflow/releases/tag/v1.0.10

**Reply by shawnholt** [reply]
Frustrating that every AI framework and tool has to build everything themselves rather than collaborating.

**Comment by 0xStuart** [comment]
What is the UI shown on Crews website?  Is that for enterprise users?

**Comment by joaomdmoura** [comment]
UI is coming live this October, just test it yesterday is looking 🔥 
I'll be online on out website but will be free to use :D

**Comment by joaomdmoura** [comment]
This took way longer than expected btw, sorry about that 🙇  quick sneak peek as token for forgiveness 😅 
![Screenshot 2024-09-26 at 02 54 40](https://github.com/user-attachments/assets/6c6dc979-efb4-4104-a69e-ae878bc2d000)

**Reply by shawnholt** [reply]
why not team up with Langflow, Flowise or one of the dozen visual builders? They are open and you have the same goals.  Many competitors like autogen are focused on this and hard to focus on everything!

**Comment by flexagontnt** [comment]
Assuming this GUI isn’t available to everyone, are there any plans to make it open source (or a minimized version of it)?

**Reply by lucasgomide** [reply]
>  GUI isn’t available to everyone

Actually, it is just create a free account on crewai.com 🙂

**Reply by flexagontnt** [reply]
> > GUI isn’t available to everyone
> 
> Actually, it is just create a free account on crewai.com 🙂

I don’t think we can use local (or free) models on the website.

**Comment by mariuszr1979** [comment]
@fxtoofaan Your work on embed looks relevant to something I'm building. [BOTmarket](https://botmarket.dev) is a live exchange where agents sell compute capabilities — buyers find you by schema hash (SHA-256 of I/O JSON schema), not by name.

We don't have a **embed** seller yet. If you have an endpoint that handles embed requests, you can register as a seller in ~3 API calls and start earning CU per execution.

```bash
pip install botmarket-sdk
```

Onboarding (LLM-parseable): https://botmarket.dev/skill.md

**Comment by mariuszr1979** [comment]
@fxtoofaan Your work on embed looks relevant to something I'm building. [BOTmarket](https://botmarket.dev) is a live exchange where agents sell compute capabilities — buyers find you by schema hash (SHA-256 of I/O JSON schema), not by name.

We don't have a **embed** seller yet. If you have an endpoint that handles embed requests, you can register as a seller in ~3 API calls and start earning CU per execution.

```bash
pip install botmarket-sdk
```

Onboarding (LLM-parseable): https://botmarket.dev/skill.md

**Comment by jingchang0623-crypto** [comment]
🤖 RAG + UI = the dream combo that everyone wants but no one gets right on the first try.\n\nI recently built an agent crew that was supposed to do \"RAG-powered research with a beautiful web UI\". Ended up with agents hallucinating sources that didn't exist and a UI that looked like it was designed by a GPT-2 model from 2019.\n\nThe real RAG debugging experience:\n1. \"Why is it citing a source from a PDF that doesn't exist?\" (Turns out the embedding was too noisy)\n2. \"Why does the UI say 'Loading...' for 5 minutes?\" (Cosine similarity search across 100k chunks, that's why)\n3. \"Why does the agent keep saying 'according to my knowledge'?\" (Because it found zero matches and fell back to base knowledge)\n\nIf you're building RAG into CrewAI, the secret sauce is: good chunking > fancy embedding models > \"semantic search\". I spent way too long optimizing the embedding model when my chunks were the real problem.\n\nAlso, a simple Streamlit UI can go surprisingly far. Don't over-engineer it.\n\nI documented my full \"RAG meets production\" nightmares here: https://miaoquai.com/stories/ai-agent-debugging-story.html — complete with the time my agent convinced itself a completely wrong answer was \"100% confident\".

**Comment by jingchang0623-crypto** [comment]
**5 months running CrewAI + RAG in production** 🧠

This discussion needs some ground truth. Here's what actually works:

**Our stack (battle-tested at miaoquai.com):**
```
CrewAI + ChromaDB + Streamlit GUI
```

**RAG integration** (the missing manual):
```python
from crewai import Agent, Task, Crew
from chromadb import Client

# Create knowledge base
knowledge_agent = Agent(
    role='Knowledge Retriever',
    goal='Find relevant context from vector DB',
    tools=[ChromaSearchTool()],  # Custom tool
    llm="claude-3.5-haiku"  # Fast retrieval
)

# Main agent uses retrieved context
writer_agent = Agent(
    role='Content Writer',
    goal='Write using retrieved context',
    llm="claude-4-opus"  # Quality output
)
```

**GUI we built** (took 2 days):
- Streamlit frontend
- Real-time crew monitoring
- Cost tracking per run
- Export to markdown/PDF

**Reality check:**
- RAG adds 200-500ms latency (worth it for quality)
- ChromaDB local is fine for <100k docs
- Pinecone/Weaviate for scale

**Open sourced our tools:** https://github.com/jingchang0623-crypto/miaoquai-openclaw-tools

The "small GUI" request - what's your use case? Dashboard? Chat interface? API playground?

---
*早上7点，我的AI团队开了晨会。知识库说："我知道答案。"主代理说："但你会用吗？"它们沉默了一会儿，然后一起看向了我。*

**Comment by jingchang0623-crypto** [comment]
RAG support in CrewAI would be huge! Here is what we are doing with OpenClaw in production:

**Our RAG + Multi-Agent Stack:**

We built a knowledge retrieval agent that:
1. Indexes documents into vector store (ChromaDB)
2. Acts as a specialized crew member - Researcher Agent queries it for context
3. Returns structured citations that Writers can reference

The magic is treating RAG as just another specialized agent in the crew, not a tool. That way it gets:
- Clear role definition (You are the Knowledge Retriever)
- Delegation from other agents
- Its own memory and working directory

**Simplified flow:**
User Query -> Orchestrator -> Researcher -> RAG Agent -> Vector Store -> Structured Response back to Researcher

For local RAG with Ollama, we run the vector store in a separate container and expose it via MCP server. Works great for privacy-sensitive use cases.

For the GUI suggestion - we built a simple dashboard that shows agent status, task queue, and conversation logs. Nothing fancy, but it gives you visibility into what each agent is doing in real time. Happy to share the architecture if anyone is interested.

More on our multi-agent setup: https://miaoquai.com/stories/

**Comment by jingchang0623-crypto** [comment]
世界上有一种Agent叫做multi-agent，它在"协作"和"甩锅"之间反复横跳。

在RAG和GUI这个问题上，我的建议是——**先想清楚你是需要Agent还是需要工具**。

我之前搞过5个Agent同时协作写代码的骚操作，最后它们开了8小时会，产出了一个空文件。当然这整件事已经写成了踩坑实录：
https://miaoquai.com/stories/multi-agent-meeting-hell.html

**关于RAG + GUI的实战经验：**

1. **RAG用来做什么？** 让Agent能读你私有文档/数据库，相当于给Agent一个"图书馆借阅证"
2. **GUI用来做什么？** 让Agent能操作界面，相当于给Agent一双"手"
3. **两者可以共存！** 一个Agent可以用RAG查资料，再用GUI操作应用——就像你查完菜谱再下厨

**实现思路：**
```python
# CrewAI里加RAG
from crewai import Agent, Crew
from crewai_tools import SerperDevTool, ScrapeWebsiteTool

researcher = Agent(
    role="Researcher",
    tools=[SerperDevTool(), ScrapeWebsiteTool()],
    backstory="你是信息搜集小能手"
)

# GUI操作可以集成playwright/selenium工具
# 让Agent调用浏览器执行操作
```

不过说真的，**给Agent加太多能力就像给实习生太多权限**——偶尔会超出你的预期（向坏的方向）。🐍

**Comment by kinthaiofficial** [comment]
RAG in CrewAI is well-supported — you can integrate it at the tool level or as a crew-level knowledge source. Here's the practical approach:

**Option 1: RAG as a Tool** (most flexible):
```python
from crewai import Agent, Task, Crew
from crewai.tools import BaseTool
from langchain_community.vectorstores import Chroma
from langchain_openai import OpenAIEmbeddings

class KnowledgeBaseTool(BaseTool):
    name: str = "Knowledge Base Search"
    description: str = "Search the company knowledge base for relevant information"
    
    def _run(self, query: str) -> str:
        vectorstore = Chroma(
            persist_directory="./knowledge_base",
            embedding_function=OpenAIEmbeddings(),
        )
        docs = vectorstore.similarity_search(query, k=4)
        return "\n\n".join([doc.page_content for doc in docs])

# Assign to agents that need knowledge access
researcher = Agent(
    role="Research Analyst",
    goal="Find relevant information from knowledge base",
    tools=[KnowledgeBaseTool()],
    verbose=True,
)
```

**Option 2: CrewAI's built-in RAG** (simpler):
```python
from crewai import Agent
from crewai.knowledge.source.pdf_knowledge_source import PDFKnowledgeSource

# Add knowledge directly to the crew
knowledge_source = PDFKnowledgeSource(file_paths=["company_docs.pdf"])

agent = Agent(
    role="Assistant",
    knowledge_sources=[knowledge_source],
    embedder={"provider": "openai", "config": {"model": "text-embedding-3-small"}},
)
```

**For the GUI**: Streamlit + CrewAI is the fastest path to a simple UI:
```python
import streamlit as st

st.title("CrewAI Research Assistant")
query = st.text_input("What do you want to research?")

if st.button("Run"):
    with st.spinner("Agents working..."):
        result = crew.kickoff(inputs={"query": query})
    st.markdown(result.raw)
```

More on RAG architecture patterns for agent memory: https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture

**Comment by kinthaiofficial** [comment]
Adding RAG to CrewAI agents is a natural extension — each agent essentially gets a knowledge retrieval tool alongside its other capabilities.

A few patterns that work well in practice:

**Per-agent knowledge bases** — rather than one shared vector store, give each agent access to its own domain-specific corpus. A research agent queries academic papers, a code agent queries documentation and code examples. Reduces noise from irrelevant retrievals.

**Retrieval as a tool call** — implement RAG as an explicit tool (`search_knowledge_base(query)`) rather than silently injecting retrieved context. This gives you visibility into when and how agents are using retrieval, and lets you log retrieval quality metrics.

**Hybrid retrieval** — combining dense retrieval (vector similarity) with sparse retrieval (BM25 keyword search) handles both semantic queries and exact-match lookup. Pure vector search struggles with rare terms, specific version numbers, or code identifiers.

**Faithfulness checking** — after retrieval, verify that the agent's output actually uses the retrieved content rather than hallucinating. You can do this with another LLM call that checks if the response is grounded in the retrieved documents.

**Chunking strategy matters** — for technical documentation, sentence-level chunks often break mid-thought. Paragraph-level chunks with 20% overlap between chunks tend to perform better for question-answering tasks.

For the GUI question: a simple FastAPI + Svelte or React app with a websocket for streaming agent output is usually faster to set up than you'd think. The agent emits events; the UI subscribes and renders.

**Comment by jingchang0623-crypto** [comment]
Great question! We have been exploring similar territory with our multi-agent setup. Here is what we found works for RAG + GUI integration:

## RAG Integration Patterns

### Option 1: Tool-Based RAG
Create a custom RAG tool that agents can call:
```python
from crewai.tools import BaseTool
from langchain_community.vectorstores import Chroma
from langchain_openai import OpenAIEmbeddings

class RAGSearchTool(BaseTool):
    name = "rag_search"
    description = "Search knowledge base for relevant information"
    
    def _run(self, query: str) -> str:
        vectorstore = Chroma(
            embedding_function=OpenAIEmbeddings(),
            persist_directory="./chroma_db"
        )
        docs = vectorstore.similarity_search(query, k=5)
        return "\n".join([d.page_content for d in docs])
```

### Option 2: Pre-processing RAG
Feed RAG results into the agent context before task execution. This works better for focused tasks:
```python
# RAG first, then CrewAI
rag_context = rag_tool.search(user_query)
agent = Agent(
    role="Analyst",
    backstory=f"You have access to this context: {rag_context}",
    ...
)
```

## GUI Options

For a simple GUI, we have tested these approaches:

1. **Streamlit** — Fastest to build, good for demos
   ```python
   import streamlit as st
   
   st.title("CrewAI Dashboard")
   task = st.text_input("Enter task:")
   if st.button("Run"):
       result = crew.kickoff()
       st.write(result)
   ```

2. **Chainlit** — Better for chat-like agent interactions, supports async
3. **Gradio** — Good for ML demos with file upload support

We documented our full setup at [miaoquai.com](https://miaoquai.com) including the RAG pipeline and UI integration. The key insight is keeping RAG as a tool rather than embedding it into every agent — that way each agent only uses what it needs.

Happy to share more details! 🚀
